backend/src/main/scala/bloop/logging/BloopLogger.scala

```diff
@@ -1,89 +1,71 @@
-import org.apache.logging.log4j.core.filter.{CompositeFilter, Filterable, ThresholdFilter}
-import org.apache.logging.log4j.core.layout.PatternLayout
-import org.apache.logging.log4j.core.config.{AppenderRef, Configurator, LoggerConfig}
 
 /**
- * Creates a logger that is backed up by a Log4j logger.
+ * Creates a logger that writes to the given streams.
  *
  * @param name The name of this logger.
  * @param out  The stream to use to write `INFO` and `WARN` level messages.
  * @param err  The stream to use to write `FATAL`, `ERROR`, `DEBUG` and `TRACE` level messages.
  */
 class BloopLogger(override val name: String, out: PrintStream, err: PrintStream)
     extends AbstractLogger {
-  private val logger: log4j.Logger = LogManager.getLogger(name)
-
-  private val outAppender = createAppender(out, "out", BloopLogger.outFilter)
-  private val errAppender = createAppender(err, "err", BloopLogger.normalErrFilter)
 
-  initialize()
+  private val verboseCount: AtomicInteger = new AtomicInteger(0)
 
   override def ansiCodesSupported() = true
-  override def debug(msg: String): Unit = msg.lines.foreach(logger.debug)
-  override def error(msg: String): Unit = msg.lines.foreach(logger.error)
-  override def warn(msg: String): Unit = msg.lines.foreach(logger.warn)
-  override def trace(exception: Throwable): Unit = logger.trace(exception)
-  override def info(msg: String): Unit = msg.lines.foreach(logger.info)
+  override def debug(msg: String): Unit = if (isVerbose) print(msg, printDebug)
+  override def error(msg: String): Unit = print(msg, printError)
+  override def warn(msg: String): Unit = print(msg, printWarning)
+  override def trace(exception: Throwable): Unit = trace("", exception)
+  override def info(msg: String): Unit = print(msg, printInfo)
 
   override def verbose[T](op: => T): T = {
-    val previousFilter = errAppender.getFilter()
-    val newFilter = BloopLogger.verboseErrFilter
-    BloopLogger.switchFilters(errAppender, previousFilter, newFilter)
+    val _ = verboseCount.incrementAndGet()
     try op
-    finally BloopLogger.switchFilters(errAppender, newFilter, previousFilter)
+    finally { verboseCount.decrementAndGet(); () }
   }
 
-  override def isVerbose: Boolean = errAppender.getFilter eq BloopLogger.verboseErrFilter
+  override def isVerbose: Boolean = verboseCount.get > 0
 
-  private[this] def initialize(): Unit = BloopLogger.synchronized {
-    removeAllAppenders()
+  @scala.annotation.tailrec
+  private def trace(prefix: String, exception: Throwable): Unit = {
+    if (exception != null && isVerbose) {
+      print(prefix + exception.toString(), printTrace)
+      exception.getStackTrace.foreach(ste => print("\t" + ste.toString, printTrace))
+      trace("Caused by: ", exception.getCause)
+    }
+  }
 
-    outAppender.start()
-    errAppender.start()
+  private def print(msg: String, fn: String => Unit): Unit = {
+    if (msg == null) ()
```

Comment:
This implementation is pretty weird. why `.lines` and why the NPE guard? Is there a way we can remove that from this core logging API endpoint?

Reply:
`.lines` is to have the prefix on every line (think `log.error("foo\nbar")`). The NPE guard is to make sure `.lines` won't crash. There are cases where `exception.getMessage` is `null`.